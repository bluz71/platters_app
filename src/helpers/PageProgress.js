import NProgress from 'nprogress'

class PageProgress {
  constructor () {
    this.loaded = false
  }

  start () {
    if (!this.loaded) {
      NProgress.start()
    }
  }

  done () {
    if (!this.loaded) {
      this.loaded = true
      NProgress.done()
    }

    return true
  }
}

export default PageProgress
