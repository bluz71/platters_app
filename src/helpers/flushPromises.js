// Details:
//   https://blog.rescale.com/testing-promise-side-effects-with-asyncawait
//   https://github.com/facebook/jest/issues/2157#issuecomment-279171856

const flushPromises = () => new Promise(resolve => setImmediate(resolve));

export default flushPromises;
