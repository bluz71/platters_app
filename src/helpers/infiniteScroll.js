// Code snippet based on:
//   http://blog.sodhanalibrary.com/2016/08/detect-when-user-scrolls-to-bottom-of.html
//
const infiniteScoll = (pagination, onPageEnd) => {
  const moreToScroll = pagination && pagination.next_page;
  if (!moreToScroll) {
    return;
  }

  const windowHeight = window.innerHeight;
  const body = document.body;
  const html = document.documentElement;
  const docHeight =
    Math.max(
      body.scrollHeight,
      body.offsetHeight,
      html.clientHeight,
      html.scrollHeight,
      html.offsetHeight
    ) - 20;
  const windowBottom = windowHeight + window.pageYOffset;

  if (windowBottom >= docHeight) {
    onPageEnd();
  }
};

export default infiniteScoll;
