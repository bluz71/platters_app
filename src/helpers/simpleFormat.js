// Code snippet provided by:
//   https://makandracards.com/makandra/1395-simple_format-helper-for-javascript
//
const simpleFormat = str => {
  str = str.replace(/\r\n?/, '\n')
  str = str.trim()

  if (str.length > 0) {
    str = str.replace(/\n\n+/g, '</p><p>')
    str = str.replace(/\n/g, '<br />')
    str = `<p>${str}</p>`
  }

  return str
}

export default simpleFormat
