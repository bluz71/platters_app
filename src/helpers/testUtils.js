import { appAuth } from '../lib/appAuth';

// Details:
//   https://blog.rescale.com/testing-promise-side-effects-with-asyncawait
//   https://github.com/facebook/jest/issues/2157#issuecomment-279171856
const flushPromises = () => new Promise(resolve => setImmediate(resolve));

const logInUser = () => (
  appAuth.logIn('eyJhbGciOiJIUzI1NiJ9.eyJ1c2VyIjo1MywibmFtZSI6ImRlbm5pc2I1NSI\
sImFkbWluIjpmYWxzZSwiZXhwIjoxNTM3ODQ4NDA0LCJpc3MiOiJwbGF0dGVycyIsImF1ZCI6InBs\
YXR0ZXJzX2FwcCJ9.ZUV69wWVXfiiMtQ2ZFPvgqvhZWwQ4WC1bI15Q9DpFCU')
);

const logOutUser = () => (
  appAuth.logOut()
);

export { flushPromises, logInUser, logOutUser };
