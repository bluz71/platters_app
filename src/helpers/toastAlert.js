import { toast } from 'react-toastify';

const toastAlert = (message) => {
  toast.error(message, { className: 'ToastAlert' });
};

export default toastAlert;
