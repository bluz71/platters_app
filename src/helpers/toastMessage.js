import { toast } from 'react-toastify';

export const toastAlert = (message) => {
  toast.error(message, { className: 'ToastAlert' });
};

export const toastNotice = (message) => {
  toast.success(message, { className: 'ToastNotice' });
};
