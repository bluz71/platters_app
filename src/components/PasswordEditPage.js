import React, { Component } from 'react'
import axios from 'axios'
import { Redirect } from 'react-router-dom'
import queryString from 'query-string'
import {
  Row,
  Col,
  PageHeader,
  Well,
  Form,
  FormGroup,
  ControlLabel,
  FormControl,
  Button
} from 'react-bootstrap'
import FontAwesome from 'react-fontawesome'
import { API_HOST } from '../config'
import { appAuth } from '../lib/appAuth'
import { toastAlert, toastNotice } from '../helpers/toastMessage'

class PasswordEditPage extends Component {
  constructor (props) {
    super(props)

    document.title = 'Platters App'

    this.state = {
      submitButtonText: 'Submit'
    }

    this.userSlug = props.match.params.user_id
    this.passwordChangeEndPoint = `${API_HOST}/api/users/passwords/${
      this.userSlug
    }/password`

    const params = queryString.parse(props.location.search)
    this.changeToken = Object.prototype.hasOwnProperty.call(params, 'token')
      ? params.token
      : ''

    // Strip the change token parameter from the URL and redirect to the shortened
    // URL.
    const url = `/users/${this.userSlug}/password/edit`
    props.history.replace(url)

    // Bind 'this' for callback functions.
    this.handleChangeSubmit = this.handleChangeSubmit.bind(this)
  }

  handleChangeSubmit (event) {
    event.preventDefault()

    this.setState({
      submitButtonText: (
        <div>
          <FontAwesome name='spinner' spin pulse /> Submitting...
        </div>
      )
    })

    const passwordChange = {
      password_change: {
        password: this.passwordInput.value,
        token: this.changeToken
      }
    }

    this.putPasswordChange(passwordChange)
  }

  putPasswordChange (passwordChange) {
    axios
      .put(this.passwordChangeEndPoint, passwordChange)
      .then(response => {
        this.setState({ submitButtonText: 'Submit' })
        appAuth.logIn(response.data.auth_token)
        toastNotice('Your password has been successfully reset')
        this.props.history.push('/')
      })
      .catch(error => {
        this.setState({ submitButtonText: 'Submit' })
        if (error.response && error.response.status === 400) {
          toastAlert(
            'Incorrect confirmation token, please retry resetting your password again'
          )
          this.props.history.push('/password/new')
        } else if (error.response && error.response.status === 406) {
          toastAlert(
            'Password could not be changed, please make sure the new password is at least 9 characters long'
          )
          this.passwordInput.value = ''
        } else {
          toastAlert('Server error, please try again later')
        }
      })
  }

  renderExplanation () {
    return <p>Your password will be reset. Choose a new password below.</p>
  }

  renderForm () {
    return (
      <Well>
        <Form horizontal onSubmit={this.handleChangeSubmit}>
          <ul className='list-group' />

          <FormGroup>
            <Col componentClass={ControlLabel} md={2}>
              Choose password
            </Col>
            <Col md={9}>
              <FormControl
                type='password'
                className='password'
                inputRef={input => (this.passwordInput = input)}
              />
            </Col>
          </FormGroup>

          <FormGroup>
            <Col mdOffset={2} md={9}>
              <Button
                type='submit'
                bsStyle='success'
                bsSize='small'
                className='submit'
              >
                {this.state.submitButtonText}
              </Button>
            </Col>
          </FormGroup>
        </Form>
      </Well>
    )
  }

  render () {
    if (this.changeToken.length !== 40) {
      toastAlert(
        'Invalid password reset token, expecting a 40 character token, please click the email link again'
      )
      return <Redirect to='/' />
    }

    return (
      <Row className='PasswordEdit'>
        <Col md={10} mdOffset={1}>
          <PageHeader>Change your password</PageHeader>
          {this.renderExplanation()}
          {this.renderForm()}
        </Col>
      </Row>
    )
  }
}

export default PasswordEditPage
