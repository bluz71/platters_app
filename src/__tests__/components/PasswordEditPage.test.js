import React from 'react'
import { shallow, mount } from 'enzyme'
import { MemoryRouter } from 'react-router-dom'
import { flushPromises } from '../../helpers/testUtils'
import * as toastModule from '../../helpers/toastMessage'
import PasswordEditPage from '../../components/PasswordEditPage'

describe('<PasswordEditPage />', () => {
  it('renders without crashing', () => {
    shallow(
      <MemoryRouter initialEntries={[{ key: 'testKey' }]}>
        <PasswordEditPage />
      </MemoryRouter>
    )
  })

  it('renders password edit form', () => {
    const match = { params: { user_id: 'fred' } }
    const location = {
      search: '?token=08a4f4f96303e9b76e60c9050520078da0621821'
    }
    const history = { replace: jest.fn() }
    const wrapper = mount(
      <MemoryRouter initialEntries={[{ key: 'testKey' }]}>
        <PasswordEditPage match={match} location={location} history={history} />
      </MemoryRouter>
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('with valid password', async () => {
    const spyNotice = jest.spyOn(toastModule, 'toastNotice')
    const match = { params: { user_id: 'fred' } }
    const location = {
      search: '?token=08a4f4f96303e9b76e60c9050520078da0621821'
    }
    const history = { replace: jest.fn() }
    const wrapper = mount(
      <MemoryRouter initialEntries={[{ key: 'testKey' }]}>
        <PasswordEditPage match={match} location={location} history={history} />
      </MemoryRouter>
    )
    wrapper.find('input.password').instance().value = 'password9'
    wrapper.find('button').simulate('submit')
    await flushPromises()
    expect(spyNotice).toHaveBeenCalledWith(
      'Your password has been successfully reset'
    )
    spyNotice.mockRestore()
  })

  it('with invalid token', async () => {
    const spyAlert = jest.spyOn(toastModule, 'toastAlert')
    const match = { params: { user_id: 'john' } }
    const location = {
      search: '?token=08a4f4f96303e9b76e60c9050520078da0621821'
    }
    const history = { replace: jest.fn(), push: jest.fn() }
    const wrapper = mount(
      <MemoryRouter initialEntries={[{ key: 'testKey' }]}>
        <PasswordEditPage match={match} location={location} history={history} />
      </MemoryRouter>
    )
    wrapper.find('input.password').instance().value = 'password9'
    wrapper.find('button').simulate('submit')
    await flushPromises()
    expect(spyAlert).toHaveBeenCalledWith(
      'Incorrect confirmation token, please retry resetting your password again'
    )
    spyAlert.mockRestore()
  })

  it('with invalid new password', async () => {
    const spyAlert = jest.spyOn(toastModule, 'toastAlert')
    const match = { params: { user_id: 'eric' } }
    const location = {
      search: '?token=08a4f4f96303e9b76e60c9050520078da0621821'
    }
    const history = { replace: jest.fn() }
    const wrapper = mount(
      <MemoryRouter initialEntries={[{ key: 'testKey' }]}>
        <PasswordEditPage match={match} location={location} history={history} />
      </MemoryRouter>
    )
    wrapper.find('input.password').instance().value = 'password'
    wrapper.find('button').simulate('submit')
    await flushPromises()
    expect(spyAlert).toHaveBeenCalledWith(
      'Password could not be changed, please make sure the new password is at least 9 characters long'
    )
    spyAlert.mockRestore()
  })
})
