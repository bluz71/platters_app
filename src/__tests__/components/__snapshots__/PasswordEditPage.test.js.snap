// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PasswordEditPage /> renders password edit form 1`] = `
<MemoryRouter
  initialEntries={
    Array [
      Object {
        "key": "testKey",
      },
    ]
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "testKey",
            "pathname": "/",
            "search": "",
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "testKey",
          "pathname": "/",
          "search": "",
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <PasswordEditPage
      history={
        Object {
          "replace": [MockFunction] {
            "calls": Array [
              Array [
                "/users/fred/password/edit",
              ],
            ],
            "results": Array [
              Object {
                "isThrow": false,
                "value": undefined,
              },
            ],
          },
        }
      }
      location={
        Object {
          "search": "?token=08a4f4f96303e9b76e60c9050520078da0621821",
        }
      }
      match={
        Object {
          "params": Object {
            "user_id": "fred",
          },
        }
      }
    >
      <Row
        bsClass="row"
        className="PasswordEdit"
        componentClass="div"
      >
        <div
          className="PasswordEdit row"
        >
          <Col
            bsClass="col"
            componentClass="div"
            md={10}
            mdOffset={1}
          >
            <div
              className="col-md-10 col-md-offset-1"
            >
              <PageHeader
                bsClass="page-header"
              >
                <div
                  className="page-header"
                >
                  <h1>
                    Change your password
                  </h1>
                </div>
              </PageHeader>
              <p>
                Your password will be reset. Choose a new password below.
              </p>
              <Well
                bsClass="well"
              >
                <div
                  className="well"
                >
                  <Form
                    bsClass="form"
                    componentClass="form"
                    horizontal={true}
                    inline={false}
                    onSubmit={[Function]}
                  >
                    <form
                      className="form-horizontal"
                      onSubmit={[Function]}
                    >
                      <ul
                        className="list-group"
                      />
                      <FormGroup
                        bsClass="form-group"
                      >
                        <div
                          className="form-group"
                        >
                          <Col
                            bsClass="col"
                            componentClass={[Function]}
                            md={2}
                          >
                            <ControlLabel
                              bsClass="control-label"
                              className="col-md-2"
                              srOnly={false}
                            >
                              <label
                                className="col-md-2 control-label"
                              >
                                Choose password
                              </label>
                            </ControlLabel>
                          </Col>
                          <Col
                            bsClass="col"
                            componentClass="div"
                            md={9}
                          >
                            <div
                              className="col-md-9"
                            >
                              <FormControl
                                bsClass="form-control"
                                className="password"
                                componentClass="input"
                                inputRef={[Function]}
                                type="password"
                              >
                                <input
                                  className="password form-control"
                                  type="password"
                                />
                              </FormControl>
                            </div>
                          </Col>
                        </div>
                      </FormGroup>
                      <FormGroup
                        bsClass="form-group"
                      >
                        <div
                          className="form-group"
                        >
                          <Col
                            bsClass="col"
                            componentClass="div"
                            md={9}
                            mdOffset={2}
                          >
                            <div
                              className="col-md-9 col-md-offset-2"
                            >
                              <Button
                                active={false}
                                block={false}
                                bsClass="btn"
                                bsSize="small"
                                bsStyle="success"
                                className="submit"
                                disabled={false}
                                type="submit"
                              >
                                <button
                                  className="submit btn btn-sm btn-success"
                                  disabled={false}
                                  type="submit"
                                >
                                  Submit
                                </button>
                              </Button>
                            </div>
                          </Col>
                        </div>
                      </FormGroup>
                    </form>
                  </Form>
                </div>
              </Well>
            </div>
          </Col>
        </div>
      </Row>
    </PasswordEditPage>
  </Router>
</MemoryRouter>
`;
